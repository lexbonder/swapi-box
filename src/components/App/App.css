* {
  box-sizing: border-box;
}

body{ 
  background-image: url('../../space.jpg');
  background-size: cover;
}