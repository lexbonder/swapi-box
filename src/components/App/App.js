import React, { Component } from 'react';
import CardContainer from '../CardContainer/CardContainer';
import Control from '../Control/Control';
import OpeningText from '../OpeningText/OpeningText';
import DataCleaner from '../../dataCleaner';
import fetchApi from '../../apiFetcher';
import './App.css';

class App extends Component {
  constructor(props) {
    super();
    this.state = {
      openingText: {},
      chosenCategory: '',
      cleanedData: [],
      favorites: [],
      numFav: 0,
    }
    this.cleaner = new DataCleaner(fetchApi)
  }

  componentDidMount = () => {
    const episodeNumber = Math.ceil(Math.random() * 7)
    this.getMovieCrawl(episodeNumber)
  }

  getMovieCrawl = async (episodeNumber) => {
    const movieInfo = await fetchApi(`https://swapi.co/api/films/${episodeNumber}/`)
    const openingText = await this.cleaner.cleanData('opening', movieInfo)
    this.setState({openingText});
  }

  toggleFavorite = (event) => {
    const clicked = event.target.name
    const clickedObject = this.state.cleanedData.find(object => object.name === clicked)
    const exists = this.state.favorites.filter(object => object.name === clicked)
    if (!exists.length) {
      this.addFavorite(clickedObject)
    } else {
      this.removeFavorite(clickedObject)
    }
  }

  addFavorite = (clickedObject) => {
    let numFav = this.state.favorites.length
    numFav++
    this.setState({favorites: [...this.state.favorites, clickedObject], numFav})
  }

  removeFavorite = (clickedObject) => {
    const remainingFavorites = this.state.favorites.filter(object => object.name !== clickedObject.name)
    let numFav = remainingFavorites.length
    this.setState({favorites: remainingFavorites, numFav})
  }

  handleClick = (event) => {
    const chosenCategory = event.target.name;
    if (chosenCategory === 'favorites') {
      this.getFavorites(chosenCategory)
    } else {
      this.getCards(chosenCategory)
    }
  }

  getFavorites = (chosenCategory) => {
    this.setState({chosenCategory, cleanedData: this.state.favorites})
  }

  getCards = async (chosenCategory) => {
    const {results} = await fetchApi(`https://swapi.co/api/${chosenCategory}/`)
    const cleanedData = await this.cleaner.cleanData(chosenCategory, results)
    this.setState({ chosenCategory, cleanedData })
  }

  render() {
    return (
      <div className="App">
        <OpeningText openingText={this.state.openingText} />
        <section className='bottom-half'>
          <Control
            handleClick={this.handleClick}
            numFav={this.state.numFav}
          />
          <CardContainer
            chosenCategory={this.state.chosenCategory}
            cleanedData={this.state.cleanedData}
            toggleFavorite={this.toggleFavorite}
            currentFavorites={this.state.favorites}
          />
        </section>
      </div>
    );
  }
}

export default App;
