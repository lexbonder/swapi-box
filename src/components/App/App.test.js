import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';
import App from './App';

describe('App', () => {  
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should match the snapshot', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  })

  it('When getCards is run it should set the name of the button in state', () => {
    const wrapper = shallow(<App />);
    // ----- is this right???? It get's the test passing..

    console.log(wrapper.state())
    // expect(wrapper.state().chosenCategory).toEqual('people')
  })


})

// // Async Tests -----


// describe('The thing', () => {
//   let mockGrocery;
//   let mockGroceries;
//   let mockUpdateGroceryList;
//   let mockEvent;

//   // Happens one time 
//   beforeAll(() => {
//       // Mock Data: 
//     mockGrocery = {
//       name: 'beef jerkeys',
//       quantity: 100
//     }

//     mockGroceries = [
//       {name: 'queso', quantity: 1000},
//       {name: 'beef jerkeys', quantity: 100}
//     ]
//       // Mock function
//     mockUpdateGroceryList = jest.fn()

//       // Mock event
//     mockEvent = { preventDefault: jest.fn() }


//     })

//   // Before each - used when you need a new instance of something (shallow/mount)
//     beforeEach(() => {
//       renderedComponent = shallow(<AddGroceryForm 
//         updateGroceryList={mockUpdateGroceryList}
//       />)
//         // Mock fetch
//       window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
//         json: () => Promise.resolve({
//           groceries.mockGroceries
//         })
//       }))
//     })
    
//   })
//   // After All - when you want to reset something back to default after all tests
//   // After Each - set conditions back to default if tests are dependant on each other
// })

// // Is fetch being called with the right parameters?
// it('calls fetch with the correct params', () => {
  
//   const expectParams = [
//     '/api/vi/groceries',
//     {
//       method: 'POST',
//       body: JSON.stringify({ grocery: mockGrocery }),
//       headers: { 'Content-Type': 'application/json' },
//     }
//   ]

//   renderedComponent.setState({grocery: mockGrocery})
//   renderedComponent.instance().handleAddGrocery(mockEvent)
//   expect(window.fetch).toHaveBeenCalledWith(...expectParams)
// })

// // Is setState returned to its default?
// it('resets the state after adding a new grocery', async () => {
//   renderedComponent.setState({ grocery: mockGrocery })

//   await renderedComponent.instance().handleAddGrocery(mockEvent)
//   await renderedComponent.update()
  
//   expect(renderedComponent.state('grocery')).toEqual({name: '', quantity: ''})
// })

// // Is the update function called?
//   // (We'll test it's functionality later, only need to know if it's called)
// it('calls the update callback', async () => {
//   await renderedComponent.instance().handleAddGrocery(mockEvent)
//   expect(mockUpdateGroceryList).toHaveBeenCalled()
// })

// // Is the error state set correctly on error?
//   // (We can force an error and test our apps behavior)
// it('updates the error state in the eent of an error', async () => {
//   window.fetch = jest.fn().mockImplementation(() => Promise.reject(new Error('oops!')))

//     await renderedComponent.instance().handleAddGrocery(mockEvent)
//     await renderedComponent.update()
//     expect(renderedComponent.state('errorState')).toEqual('oops!')
// })

