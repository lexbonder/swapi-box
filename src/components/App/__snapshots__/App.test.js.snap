// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <OpeningText
          openingText={Object {}}
/>,
        <section
          className="bottom-half"
>
          <Control
                    handleClick={[Function]}
                    numFav={0}
          />
          <CardContainer
                    chosenCategory=""
                    cleanedData={Array []}
                    currentFavorites={Array []}
                    toggleFavorite={[Function]}
          />
</section>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "openingText": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Control
              handleClick={[Function]}
              numFav={0}
/>,
            <CardContainer
              chosenCategory=""
              cleanedData={Array []}
              currentFavorites={Array []}
              toggleFavorite={[Function]}
/>,
          ],
          "className": "bottom-half",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleClick": [Function],
              "numFav": 0,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "chosenCategory": "",
              "cleanedData": Array [],
              "currentFavorites": Array [],
              "toggleFavorite": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "section",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <OpeningText
            openingText={Object {}}
/>,
          <section
            className="bottom-half"
>
            <Control
                        handleClick={[Function]}
                        numFav={0}
            />
            <CardContainer
                        chosenCategory=""
                        cleanedData={Array []}
                        currentFavorites={Array []}
                        toggleFavorite={[Function]}
            />
</section>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "openingText": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Control
                handleClick={[Function]}
                numFav={0}
/>,
              <CardContainer
                chosenCategory=""
                cleanedData={Array []}
                currentFavorites={Array []}
                toggleFavorite={[Function]}
/>,
            ],
            "className": "bottom-half",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handleClick": [Function],
                "numFav": 0,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "chosenCategory": "",
                "cleanedData": Array [],
                "currentFavorites": Array [],
                "toggleFavorite": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "section",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
